### PR TITLE
FEAT: Add support for Flexible Model

### PR DESCRIPTION
**Pull request by shellc**
Flexible Model is a customizable inference pipeline within xinference, supporting a range of tasks including image segmentation, traditional machine learning regression, and more.

Mechanism:

Create the model object with custom definitions and initialization functions.
Use xinference.model.flexible.launchers.transformers to load transformer models.
Invoke the model via the /v1/flexible/infers endpoint.
Access the model using the SDK’s infer interface.

Examples:
```
from xinference.client import Client

client = Client("http://localhost:9997")

model = client.get_model("flexible-model")

result = model.infer(inputs="happy")
```

image

### CODE REVIEW

**Comment by qinxuye** [comment]
@yiboyasss Could you review the frontend part of this PR?

File: xinference/model/flexible/core.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        """
+        raise NotImplementedError("infer method not implemented.")
+
+    @property
+    def model_uid(self):
+        return self._model_uid
+
+    @property
+    def model_path(self):
+        return self.model_path
```
Comment (qinxuye) [inline xinference/model/flexible/core.py:190]:
```suggestion
        return self._model_path
```

File: xinference/model/flexible/tests/test_flexible_models.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            model_uri="mock",
+            launcher="xinference.model.flexible.launchers.transformers",
+        ),
+        task="mock",
+    )
+
+    model.load()
+
+    result = model.infer(inputs="hello world")
+    # assert result == {"inputs": "hello world"}
```
Comment (qinxuye) [inline xinference/model/flexible/tests/test_flexible_models.py:60]: Could remove this line.

Comment (qinxuye) [reply xinference/model/flexible/tests/test_flexible_models.py:60]: And the result is not tested?

**Review by qinxuye — COMMENTED** [review]
I left some comments

File: xinference/api/restful_api.py [hunk 3]
```diff
@@ -1344,6 +1354,40 @@ async def create_variations(
...
+            raise HTTPException(status_code=500, detail=str(e))
+
+        try:
+            result = await model.infer(**kwargs)
+            return Response(result, media_type="application/json")
+        except RuntimeError as re:
+            logger.error(re, exc_info=True)
+            await self._report_error_event(model_uid, str(re))
+            self.handle_request_limit_error(re)
+            raise HTTPException(status_code=400, detail=str(re))
```
[the remark this answers is not in the document]
Comment (shellc) [reply xinference/api/restful_api.py:1385]: Maintain consistency with other APIs, such as create_embedding, create_rerank, etc.

The RuntimeError may stem from a user request that triggers a manageable runtime error. It is advisable to address this issue uniformly in future implementations.

File: xinference/model/flexible/core.py (hunk 1 above)
Comment (shellc) [reply xinference/model/flexible/core.py:190]: fix it

File: xinference/model/flexible/tests/test_flexible_models.py (hunk 2 above)
Comment (shellc) [reply xinference/model/flexible/tests/test_flexible_models.py:60]: There are Several assertions after this line.

This is a MockModel and cannot perform actual computations. The behavior of this model is to echo the inputs.
The purpose of the tests here is to evaluate the initialization and invocation of the model.

**Comment by yiboyasss** [comment]
1. Input boxes can undergo some JSON format validation to ensure that correct JSON data can be passed.
截屏2024-07-09 17 15 51


2. Add a check between lines 75 and 190 in the registerModel.js file to ensure the correct echo of the editing mode data.
截屏2024-07-09 17 17 04

**Comment by yiboyasss** [comment]
BUG: When the registered model name already exists, an error message appears when registering again, but the previously registered model with the same name will also disappear, which is not reasonable. Example: When a model with the name 'flexible-model' already exists, registering the same name again will cause the previous' flexible-model 'model to disappear.

**Review by qinxuye — APPROVED** [review]
LGTM, thanks for your contribution.

**Comment by liyingya-123** [comment]
怎么通过curl来访问这个部署呢？

**Pull request merged**